ir/visitor.cpp

```diff
@@ -20,76 +20,68 @@ limitations under the License.
 
 class Visitor::ChangeTracker {
     struct visit_info_t {
-        bool visit_finished;
+        bool visit_in_progress;
         const IR::Node *result;
     };
-    // typedef unordered_map<const IR::Node *, std::pair<bool, const IR::Node *>>  visited_t;
     typedef unordered_map<const IR::Node *, visit_info_t>  visited_t;
     visited_t           visited;
 
  public:
-    struct change_t {
-        // std::pair<visited_t::iterator, bool>    state;  // result of visited.emplace
-        bool           valid;
-        bool           already_present;
-        const IR::Node *original_node;
-        visit_info_t   *visit_info;    // Points into ChangeTracker.visited.
-
-        change_t() : valid(false) {}
-        change_t(visited_t *visited, const IR::Node *n) : valid(true) {
-                // Initialization
-                visited_t::iterator visited_it;
-                bool inserted;
-                std::tie(visited_it, inserted) =
-                    visited->emplace(n, visit_info_t({false,n}));
-                already_present = !inserted;
-                original_node = visited_it->first;
-                visit_info = &(visited_it->second);
-
-                // Sanity check for IR loops
-                if (!inserted && !visit_info->visit_finished)
-                    BUG("IR loop detected "); }
-
-        explicit operator bool() { return valid; }
-        bool done() { return already_present; }
-        const IR::Node *orig() { return original_node; }
-        const IR::Node *result() { return visit_info->result; }
-    };
 
-    bool done(const IR::Node *n) const {
-        return visited.count(n) && visited.at(n).visit_finished;
+    /* Begin tracking `n` during a visiting pass.  Use `finish` to mark `n` as
+     * visited once the pass completes.
+    */
+    void start(const IR::Node *n) {
+        // Initialization
+        visited_t::iterator visited_it;
+        bool inserted;
+        std::tie(visited_it, inserted) =
+            visited.emplace(n, visit_info_t({true,n}));
+
+        // Sanity check for IR loops
+        bool already_present = !inserted;
+        visit_info_t *visit_info = &(visited_it->second);
+        if (already_present && visit_info->visit_in_progress)
+            BUG("IR loop detected "); 
     }
-    const IR::Node *result(IR::Node *n) const { return visited.at(n).result; }
 
-    change_t track(const IR::Node *n) { return change_t(&visited, n); }
-    void start(change_t &change) { change.visit_info->visit_finished = false; }
-    bool finish(change_t &change, const IR::Node *orig, const IR::Node *final) {
+    /* Mark `n` as finished.  `done(n)` will return true, and `result(n)` will
```

Comment:
It'd be good to make this a Doxygen comment as well. For short documentations comments like this, you can also use this syntax:

```
/// Mark `n` as finished.  `done(n)` will return true, and `result(n)` will
/// return the resulting node, if any.
```

Even better would be to refer to the actual parameter names. Presumably by `n` you mean `orig`? In Doxygen, you refer to a parameter using`@`, so you might start with something like this:

```
/// Mark @orig as finished.  `done(@orig)` will return true, and `result(@orig)` will
```

It'd be good to document what `@final` is used for, and what the return value is. In Doxygen, you refer to the return value (which doesn't have a name, obviously) with `@return`. So perhaps you might use something along these lines? (You understand this code better than me, so I'm sure you can write a clearer explanation.)

```
/**
 * Mark the process of visiting @orig as finished, with @final being the final state
 * of the node, or nullptr if the node was removed from the tree. `done(@orig)`
 * will return true, and `result(@orig)` will return the resulting node, if any.
 *
 * @return true if the node has changed or been removed.
 */
```